test/opt/eliminate_dead_functions_test.cpp

```diff
@@ -0,0 +1,189 @@
+          "%7 = OpTypeFunction %void",
+      "%float = OpTypeFloat 32",
+          "%9 = OpTypeFunction %float",
+  "%float_0_5 = OpConstant %float 0.5",
+  "%float_0_8 = OpConstant %float 0.8",
+    "%v4float = OpTypeVector %float 4",
+"%_ptr_Output_v4float = OpTypePointer Output %v4float",
+"%gl_FragColor = OpVariable %_ptr_Output_v4float Output",
+    "%float_1 = OpConstant %float 1",
+       "%main = OpFunction %void None %7",
+         "%15 = OpLabel",
+         "%16 = OpFunctionCall %float %Constant",
+         "%17 = OpFunctionCall %float %Constant",
+         "%18 = OpCompositeConstruct %v4float %16 %17 %float_0_8 %float_1",
+               "OpStore %gl_FragColor %18",
+               "OpReturn",
+               "OpFunctionEnd",
+  "%Constant = OpFunction %float None %9",
+         "%20 = OpLabel",
+               "OpReturnValue %float_0_8",
+               "OpFunctionEnd"
+      // clang-format on
+  };
+
+  const std::vector<const char*> dead_function = {
+      // clang-format off
+      "%Dead = OpFunction %float None %9",
+         "%19 = OpLabel",
+               "OpReturnValue %float_0_5",
+               "OpFunctionEnd",
+      // clang-format on
+  };
+
+  SetAssembleOptions(SPV_TEXT_TO_BINARY_OPTION_PRESERVE_NUMERIC_IDS);
+  SinglePassRunAndCheck<opt::EliminateDeadFunctionsPass>(
+      JoinAllInsts(Concat(common_code, dead_function)),
+      JoinAllInsts(common_code), /* skip_nop = */ true);
+}
+
+TEST_F(EliminateDeadFunctionsBasicTest, BasicKeepLiveFunction) {
+  // Everything is reachable from an entry point, so no functions should be
+  // deleted.
+  const std::vector<const char*> text = {
+      // clang-format off
+               "OpCapability Shader",
+               "OpMemoryModel Logical GLSL450",
+               "OpEntryPoint Fragment %main \"main\" %gl_FragColor",
+               "OpExecutionMode %main OriginUpperLeft",
+               "OpSource GLSL 150",
+               "OpName %main \"main\"",
+               "OpName %Dead \"Dead\"",
+               "OpName %Constant \"Constant\"",
+               "OpName %gl_FragColor \"gl_FragColor\"",
+               "OpDecorate %gl_FragColor Location 0",
+       "%void = OpTypeVoid",
+          "%7 = OpTypeFunction %void",
+      "%float = OpTypeFloat 32",
+          "%9 = OpTypeFunction %float",
+  "%float_0_5 = OpConstant %float 0.5",
+  "%float_0_8 = OpConstant %float 0.8",
+    "%v4float = OpTypeVector %float 4",
+"%_ptr_Output_v4float = OpTypePointer Output %v4float",
+"%gl_FragColor = OpVariable %_ptr_Output_v4float Output",
+    "%float_1 = OpConstant %float 1",
+       "%main = OpFunction %void None %7",
+         "%15 = OpLabel",
+         "%16 = OpFunctionCall %float %Constant",
+         "%17 = OpFunctionCall %float %Dead",
+         "%18 = OpCompositeConstruct %v4float %16 %17 %float_0_8 %float_1",
+               "OpStore %gl_FragColor %18",
+               "OpReturn",
+               "OpFunctionEnd",
+      "%Dead = OpFunction %float None %9",
```

Comment:
It's misleading to call the %Dead function "Dead".  :-)

Reply:
Either this was is now out of date, or I was wrong.  :-)

Reply:
Out of date, I changed it after your comments were made.